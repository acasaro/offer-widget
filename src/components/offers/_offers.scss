@import '../../public/stylesheets/partials/mixins';

.offers {
	box-sizing: border-box;
	.offer-card {
		display: flex;
		flex-wrap: nowrap;
		border-bottom: 1px solid #d8d8d8;
		margin-top: 0;
		flex-direction: column;

		@include bp(md) {
			flex-direction: row;
			justify-content: space-between;
			align-items: center;
		}

		&__title {
			font-size: 18px;
			font-weight: bold;
			margin: 0 0 20px 0;
			@include bp(md) {
				margin: 10px 0;
			}
		}

		&__description {
			margin: 0 0 20px 0;
			@include bp(md) {
				margin: 10px 0 0 0;
			}
		}

		&__text {
			flex: 1 0 auto;

			@include bp(md) {
				flex: 1 10 auto;
			}

			&.centered {
				text-align: center;
				@include bp(md) {
					text-align: left;
				}
			}
		}

		.offer-image {
			max-width: 115px;
			margin-right: 20px;
		}

		.flex-content {
			align-self: center;
		}

		.button-col {
			display: flex;
			flex-direction: column;
			justify-content: center;
			flex: 0 0 auto;

			@include bp(md) {
				margin-left: 20px;
			}

			.button {
				display: block;
				width: 100%;
				&:first-child {
					margin-bottom: 10px;
				}
			}
		}

		.icon-container {
			align-self: flex-start;
			margin-bottom: 20px;

			@include bp(md) {
				align-self: center;
				margin: 0 20px 0 0;
			}

			&.centered {
				align-self: center;
			}
		}

		.icon-circle {
			background-color: orange;
			width: 75px;
			height: 75px;
			border-radius: 50%;
			display: table;
			text-align: center;

			.icon {
				color: #fff;
				display: table-cell;
				vertical-align: middle;
				font-size: 2.5rem;
			}
		}
	}
}
