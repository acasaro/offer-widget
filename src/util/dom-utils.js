import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { get } from 'lodash';
import { isNullEmptyOrUndefined } from './string-utils';

export const checkIfElementExists = selector => {
	const element = document.querySelector(selector);
	if (element !== null) {
		return true;
	}
	return false;
};

export const renderIfExists = (Component, selector, opts = {}) => {
	if (checkIfElementExists(selector)) {
		render(Component, document.querySelector(selector));
		return true;
	}
	// If mounting was unsuccessful, return false and put up a console warning
	console.warn(`Attempting to mount to ${selector} failed. DOM node not found.`);
	return false;
};

/**
 * linkTarget  - takes in boolean, returns '_blank' for true, '_self' for false
 * @param {Boolean}  openNewWindow       Whether to open in new tab or not
 * @returns {String}
 */
export const linkTarget = openNewWindow => (openNewWindow ? '_blank' : '_self');

export const getTabbables = element => {
	const tabbableQuery = ['input', 'select', 'a[href]', 'textarea', 'button', '[role=button]', '[tabindex]'];

	return element.querySelectorAll(tabbableQuery.join(','));
};
