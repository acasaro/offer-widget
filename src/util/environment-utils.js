import { get } from 'lodash';
import { API, APP, ENDPOINTS } from './constants/environment';

/**
 * getEnvironment - Returns the current environment, or development by default
 * @returns {String}
 */
export const getEnvironment = () => {
	if (process.env.NODE_ENV) {
		return process.env.NODE_ENV;
	}

	return 'development';
};

/**
 * userLayer7 - Whether or not to use Layer7 protected endpoints
 * @returns {Boolean}
 */
export const useLayer7 = () => {
	const environment = getEnvironment();
	return environment === 'stage' || environment === 'production';
};

/**
 * getApiUrl  - Returns the URL for the api, given the current environment
 * @returns {String}
 */
export const getApiUrl = () => {
	switch (getEnvironment()) {
		case 'production':
			return API.PRODUCTION;
		case 'stage':
			return API.STAGE;
		case 'test':
			return API.TEST;
		case 'development':
		default:
			return API.DEV;
	}
};

/**
 * getAppUrl  - Returns the URL for the app, given the environment
 * @returns {String}
 */
export const getAppUrl = () => {
	switch (getEnvironment()) {
		case 'production':
			return APP.PRODUCTION;
		case 'stage':
			return APP.STAGE;
		case 'test':
			return APP.TEST;
		case 'development':
		default:
			return APP.DEV;
	}
};

/**
 * getEndpoint  - Returns the endpoint based on environment
 * @returns {String}
 */
export const getEndpoint = service => get(ENDPOINTS, `${service}.${useLayer7() ? 'layer7' : 'insecure'}`);

/**
 * getMockData  - Returns the filepath for given mock data file
 * @param {String} fileName name of file to find path for
 * @returns {String}
 */
export const getMockData = fileName => {
	const isLocal = window.location.href.includes('localhost');
	return isLocal
		? `/mock-data/${fileName}`
		: `/etc/designs/digitaladvocacy/clientlibs/homepage-ui/mock-data/${fileName}`;
};
