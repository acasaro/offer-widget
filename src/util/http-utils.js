import axios from 'axios';
import { get as lodashGet } from 'lodash';

/**
 * getQueryString - Get a query string parameter by name
 * Adapted from https://stackoverflow.com/questions/901115/how-can-i-get-query-string-values-in-javascript
 *
 * @param {String} name   Name of parameter to retrieve
 * @returns {{String|Null}}
 */
export const getQueryString = name => {
	const url = lodashGet(window, 'location.search');
	name = name.replace(/[[\]]/g, '\\$&');
	const regex = new RegExp(`[?&]${name}(=([^&#]*)|&|#|$)`);
	const results = regex.exec(url);

	if (!results) {
		return null;
	}

	if (!results[2]) {
		return '';
	}

	return decodeURIComponent(results[2].replace(/\+/g, ' '));
};
