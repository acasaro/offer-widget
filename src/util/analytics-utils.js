import { get } from 'lodash';
import { getEnvironment } from './environment-utils';
import { POST_TYPES, SECTION_NAMES, ANALYTICS_SCRIPT_SRC } from './constants/analytics';

// To do: strip out unnecessary stuff, only post offers. to the postDataObject, and on click.

export class AnalyticsQueue {
	// Wait a max of ten seconds
	static MAX_WAIT = 10000;

	// Check every second for publishPostPageData
	static CHECK_EVERY = 1000;

	constructor() {
		this.queue = [];
		this.runningTime = 0;
	}

	// Executes publishPostPageData immediately if possible,
	// else adds analytics data to queue for retry
	queueOrExecute = (type, content) => {
		if (window.publishPostPageData) {
			window.publishPostPageData(type, content);
		} else {
			this.queue.push({ type, content });
			// If the analytics queue hasn't started processing, start it
			if (this.runningTime === 0) {
				this.executeQueue();
			}
		}
	};

	executeQueue = () => {
		// If publishPostPageData is available, send all queued
		// items to Adobe analytics, then clear the queue
		if (window.publishPostPageData) {
			this.queue.forEach(item => {
				window.publishPostPageData(item.type, item.content);
			});

			this.queue = [];
		} else if (this.runningTime < AnalyticsQueue.MAX_WAIT) {
			this.runningTime += AnalyticsQueue.CHECK_EVERY;
			window.setTimeout(this.executeQueue, AnalyticsQueue.CHECK_EVERY);
		}
	};
}

// Instantiate analytics queue immediately, but don't start timers
const analyticsQueue = new AnalyticsQueue();

// TODO: make businessUnit & website parameters, rather than hardcoded

export const updatePageDataLayer = (pageName, loginStatus, userName, referringSite, referringSiteSection) => {
	const updatedPageDataLayer = {
		content: {
			pageName,
			referringSite,
			referringSiteSection,
			siteSectionL1: '',
			siteSectionL2: '',
			siteSectionL3: '',
			businessUnit: 'optum',
			website: 'myoptum',
		},
		user: {
			loginStatus,
			userType: 'consumer',
		},
		dacoData: {
			protected: {
				user: {
					userName,
					userType: 'consumer',
					login: loginStatus,
				},
			},
		},
	};

	window.pageDataLayer = { ...window.pageDataLayer, ...updatedPageDataLayer };
};

export const updateTracking = eligibility => {
	const href = window.location.href;
	const referringSiteSection = '';
	let loginStatus = '';
	let userName = '';
	let referringSite = '';
	loginStatus = 'not loggedin';
	let page = window.location.pathname.substring(window.location.pathname.lastIndexOf('/') + 1);
	let pageName = page.replace('.html', '');

	updatePageDataLayer(pageName, loginStatus, userName, referringSite, referringSiteSection);
};

/**
 * injectAnalyticsScript  - Appends environment-specific satellite library script to document head, calls _satellite.pageBottom()
 */
export const injectAnalyticsScript = async () => {
	const satelliteLibScript = document.createElement('script');
	satelliteLibScript.async = true;
	satelliteLibScript.id = 'satLibScript';
	satelliteLibScript.src = getEnvironment() === 'production' ? ANALYTICS_SCRIPT_SRC.PROD : ANALYTICS_SCRIPT_SRC.DEFAULT;

	document.head.appendChild(satelliteLibScript);

	// Page needs analytics. Wait up to 2 seconds for analytics to have loaded
	// before bailing
	const CHECK_EVERY = 50;
	const MAX_WAIT = 2000;
	let timer = 0;
	return new Promise(resolve => {
		const loadInterval = window.setInterval(() => {
			if (window._satellite && window._satellite.pageBottom) {
				window.clearInterval(loadInterval);

				// If no pageDataLayer is present, create one
				if (!window.pageDataLayer) {
					updateTracking();
				}
				// If no publishPostPageData available, initialize the analytics script
				if (!window.publishPostPageData) {
					window._satellite.pageBottom();
				}
				resolve(true);
			} else if (timer > MAX_WAIT) {
				window.clearInterval(loadInterval);
				resolve(false);
			} else {
				timer += CHECK_EVERY;
			}
		}, CHECK_EVERY);
	});
};

export const postAnalyticsData = (type, content) => {
	const postObject = {
		user: get(window, 'pageDataLayer.user'),
		content: content,
	};

	// Either execute or add the current analytics post to the queue
	analyticsQueue.queueOrExecute(type, postObject);

	return analyticsQueue.queue.length === 0;
};
