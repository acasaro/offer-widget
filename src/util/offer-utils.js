import { get, forOwn } from 'lodash';
import { isNullEmptyOrUndefined } from './string-utils';
import { getApiUrl, getEndpoint, getMockData } from './environment-utils';
import { useMockData } from '../initialization-helper';

export const sortOffers = (offers = []) =>
	offers.sort((a, b) => {
		// Prioritize AEM sort order if applicable, otherwise use OGM-specified sort order
		const aOrder = get(a, 'content.order', a.order);
		const bOrder = get(b, 'content.order', b.order);

		return parseInt(aOrder, 10) - parseInt(bOrder, 10);
	});

/**
 * hasModal - checks if object has non-null keys
 * @param {Object} modalArray
 * @returns {Boolean}
 */
export const hasModal = modalArray => {
	// default false
	let hasModal = false;
	forOwn(modalArray, (value, key) => {
		// if any of the modal keys are not null, return true
		if (!isNullEmptyOrUndefined(value)) {
			hasModal = true;
		}
	});

	return hasModal;
};

/**
 * getOffersUrl - should return offers GET url or mock data, depending on mock data
 */

// TODO: parameterize the query strings depending on the portal making the request
export const getOffersUrl = () => {
	return useMockData()
		? getMockData('offers-mock-data.json')
		: getApiUrl() + '/' + getEndpoint('leadInteraction') + '?portal=homepage&language=en';
};
