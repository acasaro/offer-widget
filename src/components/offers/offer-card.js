import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { forOwn } from 'lodash';
import Card from '../general/card/card';
import { hasModal } from '../../util/offer-utils';
import { linkTarget } from '../../util/dom-utils';
import AnalyticsLink from '../general/analytics-link/analytics-link';
import { POST_TYPES } from '../../util/constants/analytics';
import { isNullEmptyOrUndefined } from '../../util/string-utils';
import Modal from '../general/modal/modal';

class OfferCard extends Component {
	static propTypes = {
		title: PropTypes.string,
		description: PropTypes.string,
		type: PropTypes.string,
		handleClick: PropTypes.func,
		tellMeMore: PropTypes.shape({
			title: PropTypes.string,
			link: PropTypes.string,
			openNewWindow: PropTypes.bool,
			linkIcon: PropTypes.string,
		}),
		notInterested: PropTypes.shape({
			title: PropTypes.string,
			linkIcon: PropTypes.string,
			link: PropTypes.string,
			openNewWindow: PropTypes.bool,
		}),
		icon: PropTypes.string,
		image: PropTypes.string,
		altText: PropTypes.string,
		modal: PropTypes.shape({
			title: PropTypes.string,
			content: PropTypes.string,
			cta: PropTypes.string,
			ctaLink: PropTypes.string,
			dismiss: PropTypes.string,
		}),
		portal: PropTypes.string,
	};

	static defaultProps = {
		title: '',
		description: '',
		type: '',
		tellMeMore: {},
		notInterested: {},
		icon: '',
		image: '',
		altText: '',
		modal: {},
	};

	state = {
		modalOpen: false,
	};

	// Toggles the modal (opened === true, closed === false)
	toggleModal = () => {
		// if modal is open, then send interaction
		if (this.state.modalOpen) {
			this.tellMore();
		}
		this.setState(prevState => ({ modalOpen: !prevState.modalOpen }));
	};

	//isMobile = () => window.innerWidth < 200;

	tellMore = () => this.props.handleClick('TellMeMore');

	removeOffer = () => this.props.handleClick('NoInterest');

	render = () => {
		const { title, description, type, image, icon, tellMeMore, notInterested, altText, modal, portal } = this.props;

		const { modalOpen } = this.state;

		const isCentered = Boolean(icon);
		// const isMobile = {isMobile};
		// console.log("isMobile"+isMobile);
		// console.log("window.innerWidth < 200"+window.innerWidth);

		const tellMoreClickHandler = hasModal(modal) ? this.toggleModal : this.tellMore;
		return (
			<Card extraClasses="offer-card">
				{icon && (
					<div className="icon-container icon-circle centered">
						<span className={`icon ${icon} icon-large`} alt={altText} />
					</div>
				)}

				{image && (
					<div className="icon-container">
						<img src={image} className="offer-image icon-img" alt={altText} />
					</div>
				)}
				<div className={`offer-card__text ${isCentered ? 'centered' : ''}`}>
					<h6 className="offer-card__title">{title}</h6>
					<p className="offer-card__description">{description}</p>
				</div>
				<div className="button-col">
					<AnalyticsLink
						className={`button button-primary offer-cta ${tellMeMore.linkIcon}`}
						onClick={tellMoreClickHandler}
						type={POST_TYPES.OFFER_LINKS}
						analyticsData={{
							offerName: title,
							offerType: type,
							linkName: tellMeMore.title,
							referringPortal: portal,
						}}
						href={tellMeMore.link}
						target={linkTarget(tellMeMore.openNewWindow)}
					>
						{tellMeMore.title}
					</AnalyticsLink>

					<AnalyticsLink
						className={`button button-outline remove-offer ${notInterested.linkIcon}`}
						onClick={this.removeOffer}
						type={POST_TYPES.OFFER_LINKS}
						href={notInterested.link}
						role="button"
						analyticsData={{
							offerName: title,
							offerType: type,
							linkName: notInterested.title,
							referringPortal: portal,
						}}
					>
						{notInterested.title}
					</AnalyticsLink>
				</div>

				{hasModal && (
					<Modal
						analyticsOfferName={modal.title}
						analyticsOfferType={this.props.type}
						analyticsLinkName="offer modal close"
						portal={this.props.portal}
						handleClose={this.toggleModal}
						heading={modal.title}
						isOpen={modalOpen}
						closeText={modal.dismiss}
					>
						<p>{modal.content}</p>
						<a href={'tel:' + modal.content}>{modal.content}</a>
					</Modal>
				)}
			</Card>
		);
	};
}

export default OfferCard;
