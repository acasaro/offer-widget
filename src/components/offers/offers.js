import React, { Component } from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { get } from 'lodash';
import { getApiUrl, getEndpoint, getMockData } from '../../util/environment-utils';
import { useMockData } from '../../initialization-helper';
import { postAnalyticsData } from '../../util/analytics-utils';
import { getOffersUrl } from '../../util/offer-utils';
import OfferCard from './offer-card';
import Pagination from '../general/pagination/pagination';
import './_offers.scss';
import { POST_TYPES } from '../../util/constants/analytics';

export class Offers extends Component {
	static propTypes = {
		cacheId: PropTypes.string,
		portal: PropTypes.string,
	};

	static defaultProps = {
		cacheId: '',
	};

	state = {
		offers: [],
		options: {},
		currentOffer: {},
	};

	getOffersData = () => {
		return new Promise(resolve => {
			resolve(
				axios
					.get(getOffersUrl(), {
						headers: {
							'Content-Type': 'application/json',
							UUID: this.props.cacheId,
						},
					})
					.then(this.setOffersData)
					.catch(this.handleError)
			);
		});
	};

	handleError = error => {
		if (error.response) {
			// analytics
			const postObject = {
				errorType: error.response.status,
				errorName: error.response.data.message ? error.response.data.message : error.response.statusText,
				referringPortal: this.props.portal,
			};
			postAnalyticsData(POST_TYPES.OFFER_UI_ERRORS, postObject);
		}
		return false;
	};

	setOffersData = res => {
		// only set offers data if there's at least one
		if (get(res, 'data.offers[0]')) {
			this.setState({
				offers: res.data.offers,
				currentOffer: res.data.offers[0],
				options: res.data.options,
			});
			return true;
		}
		return false;
	};

	// Fetch offers data
	componentDidMount = async () => {
		await this.getOffersData();

		// After offers data comes back, dispatch 'offersLoaded' event
		this.dispatchOffersLoaded();

		//analytics
		let objOffer = Object.keys(this.state.offers).map(key => this.state.offers[key]);
		let eligibleOffers = objOffer.map(offer => offer.title).join('|');
		const postObject = {
			offerName: this.state.currentOffer.title,
			offerType: this.state.currentOffer.type,
			eligibleOffers: eligibleOffers,
			referringPortal: this.props.portal,
		};
		postAnalyticsData(POST_TYPES.OFFER_PAGELOAD, postObject);
	};

	handleClick = interaction => {
		const offersArray = this.state.offers;
		for (let i = 0; i < this.state.offers.length; i++) {
			if (this.state.currentOffer.offerCode === this.state.offers[i].offerCode) {
				offersArray.splice(i, 1);
			}
		}
		this.setState({
			offers: offersArray,
			currentOffer: offersArray[0],
		});
		this.sendInteraction(interaction);
	};

	sendInteraction = interaction => {
		const interactionData = this.constructInteraction(interaction);
		axios
			.post(interactionData.url, interactionData.data)
			.then(this.checkActionSuccess)
			.catch(this.handleError);
	};

	checkActionSuccess = response => {
		//analytics
		const result = response.data.postResult.postOutput.status.messages.message[0].code;
		if (result === '10200') {
			const postObject = {
				status: 'success',
				referringPortal: this.props.portal,
			};
			postAnalyticsData(POST_TYPES.OFFER_API, postObject);
			return true;
		}
		return false;
	};

	constructInteraction = interaction => {
		const thisOffer = this.state.currentOffer;
		const nextOffer = this.state.offers[0];

		let dispTxt = '';
		if (interaction === 'NoInterest') {
			dispTxt = thisOffer.template.notInterested.title;

			//analytics:
			if (this.state.offers.length > 0) {
				const postObject = {
					offerName: nextOffer.title,
					offerType: nextOffer.type,
					linkName: nextOffer.template.notInterested.title,
					offerNavType: 'not interested click',
					referringPortal: this.props.portal,
				};
				postAnalyticsData(POST_TYPES.OFFER, postObject);
			}
		} else if (interaction === 'TellMeMore') {
			dispTxt = thisOffer.template.tellMeMore.title;

			//analytics:
			if (this.state.offers.length > 0) {
				const postObject = {
					offerName: nextOffer.title,
					offerType: nextOffer.type,
					linkName: nextOffer.template.tellMeMore.title,
					offerNavType: 'view details click',
					referringPortal: this.props.portal,
				};
				postAnalyticsData(POST_TYPES.OFFER, postObject);
			}
		}
		return {
			url: getApiUrl() + '/' + getEndpoint('leadPromotion') + '/' + thisOffer.offerCode,
			data: {
				blackListDays: thisOffer.blacklistTime,
				dispositionText: dispTxt,
				eventDesc: thisOffer.template.title,
				eventName: interaction,
				memberId: thisOffer.memberId,
				opportunityTypeCd: thisOffer.opportunityTypeCd,
				treatmentCode: thisOffer.treatmentCode,
			},
		};
	};

	dispatchOffersLoaded = () => {
		var event = document.createEvent('Event');
		event.initEvent('myoptumOffersLoaded', false, true);
		const isOffer = this.state.offers.length > 0 ? true : false;
		event.detail = { hasOffer: isOffer };
		window.dispatchEvent(event);
		return true;
	};

	/**
	 * handlePageChange - Changes the page to the offer with a matching offerCode\
	 *
	 * @param {String} newPage offerCode of offer being paged to
	 */
	handlePageChange = newPage => {
		const currentPageKey = this.state.currentOffer.offerCode;
		const newPageKey = newPage.offerCode;

		// If page isn't new, do nothing
		//if (currentPageKey !== newPageKey) {
		const postObject = {
			offerName: newPage.template.title,
			offerType: newPage.type,
			offerNavType: 'navigation dot click',
			referringPortal: this.props.portal,
		};
		postAnalyticsData(POST_TYPES.OFFER, postObject);
		this.setState({ currentOffer: newPage });
		//}
	};

	render = () => {
		const { offers, options, currentOffer } = this.state;
		const shouldShow = this.state.currentOffer && Object.keys(this.state.currentOffer).length > 0;

		return (
			<div className={classNames('offers', { 'has-offers': shouldShow })}>
				{shouldShow && (
					<OfferCard
						handleClick={this.handleClick}
						title={get(currentOffer, 'template.title')}
						description={get(currentOffer, 'template.content')}
						image={get(currentOffer, 'template.image')}
						icon={get(currentOffer, 'template.icon')}
						notInterested={get(currentOffer, 'template.notInterested')}
						tellMeMore={get(currentOffer, 'template.tellMeMore')}
						altText={get(currentOffer, 'template.alt')}
						modal={get(currentOffer, 'template.modal')}
						type={currentOffer.type}
						offerCode={currentOffer.offerCode}
						category={currentOffer.category}
						portal={this.props.portal}
					/>
				)}

				{offers &&
					offers.length > 1 &&
					options &&
					options.displayPagination && (
						<Pagination
							uniqueId="offerCode"
							activePage={currentOffer.offerCode}
							pages={offers}
							handlePageChange={this.handlePageChange}
						/>
					)}
			</div>
		);
	};
}

export default Offers;
