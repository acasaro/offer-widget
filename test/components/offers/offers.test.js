import React from 'react';
import axios from 'axios';
import { mount, shallow } from 'enzyme';
import { Offers, mapStateToProps } from '../../../src/components/offers/offers';
import { getApiUrl, getEndpoint } from '../../../src/util/environment-utils';
import mockOffersData from '../../../src/mock-data/offers-mock-data.json';
import { empty } from 'rxjs/observable/empty';
import { postAnalyticsData } from '../../../src/util/analytics-utils';

describe('Offers Component', () => {
	it('should render a blank offers if no props are passed in', () => {
		const offers = shallow(<Offers />);

		expect(offers).toMatchSnapshot();
	});

	it('should change page if handlePageChange is passed new page, stay same if passed same', () => {
		const offers = mount(<Offers />);
		offers.setState({ offers: mockOffersData.offers, currentOffer: mockOffersData.offers[0] });

		offers.instance().handlePageChange(mockOffersData.offers[0]);
		expect(offers.state().currentOffer).toBe(mockOffersData.offers[0]);

		offers.instance().handlePageChange(mockOffersData.offers[1]);
		expect(offers.state().currentOffer).toBe(mockOffersData.offers[1]);
	});

	it('should set offers data with setOffersData', () => {
		const offers = shallow(<Offers />);
		const emptyState = {
			offers: [],
			options: {},
			currentOffer: {},
		};

		expect(offers.state()).toEqual(emptyState);
		offers.instance().setOffersData({ data: mockOffersData });
		expect(offers.state()).toEqual({
			offers: mockOffersData.offers,
			currentOffer: mockOffersData.offers[0],
			options: mockOffersData.options,
		});
	});

	it('should return false with handleError', () => {
		const offers = shallow(<Offers />);
		let fn = () => {
			throw new Error(false);
		};
		const result = offers.instance().handleError(fn);
		expect(result).toBe(false);
	});

	it('should return true with checkActionSuccess when response code is equal to 10200', () => {
		const mockResponse = {
			data: {
				postResult: {
					postOutput: {
						resultMeta: { transactionId: '0' },
						status: { messages: { message: [{ code: '10200', name: 'OK', severity: 'INFO', description: 'null' }] } },
					},
				},
			},
		};
		const offers = shallow(<Offers />);

		const actionSuccess = offers.instance().checkActionSuccess(mockResponse);
		expect(actionSuccess).toBe(true);

		mockResponse.data.postResult.postOutput.status.messages.message[0].code = '11111';
		const actionFail = offers.instance().checkActionSuccess(mockResponse);
		expect(actionFail).toBe(false);
	});

	it('should return the correct data structure from constructInteraction', () => {
		const offers = mount(<Offers />);
		offers.setState({ offers: mockOffersData.offers, currentOffer: mockOffersData.offers[0] });
		const currentOffer = offers.state('currentOffer');

		const expectedInteraction = {
			url: getApiUrl() + '/' + getEndpoint('leadPromotion') + '/' + currentOffer.offerCode,
			data: {
				blackListDays: currentOffer.blacklistTime,
				dispositionText: currentOffer.template.notInterested.title,
				eventDesc: currentOffer.template.title,
				eventName: 'NoInterest',
				memberId: currentOffer.memberId,
				opportunityTypeCd: currentOffer.opportunityTypeCd,
				treatmentCode: currentOffer.treatmentCode,
			},
		};

		const noInterestInteraction = offers.instance().constructInteraction('NoInterest');
		expect(noInterestInteraction).toEqual(expectedInteraction);

		// Update to expect TellMeMore
		expectedInteraction.data.eventName = 'TellMeMore';
		expectedInteraction.data.dispositionText = currentOffer.template.tellMeMore.title;

		const tellMoreInteraction = offers.instance().constructInteraction('TellMeMore');
		expect(tellMoreInteraction).toEqual(expectedInteraction);
	});

	it('should send a post request through axios when sendInteraction is called', () => {
		const offers = mount(<Offers />);
		offers.setState({ offers: mockOffersData.offers, currentOffer: mockOffersData.offers[0] });

		// Construct interaction is already tested, so just test that post request is made
		const spy = jest.spyOn(axios, 'post');

		const request = offers.instance().sendInteraction('Something');
		expect(axios.post).toHaveBeenCalled();
	});

	it('should dispatch the myoptumOffersLoaded event when offers have successfully loaded', () => {
		let eventDispatched;
		window.addEventListener('myoptumOffersLoaded', () => {
			eventDispatched = true;
		});

		const offers = mount(<Offers />);
		offers.instance().dispatchOffersLoaded();

		expect(eventDispatched).toBe(true);
	});
});
