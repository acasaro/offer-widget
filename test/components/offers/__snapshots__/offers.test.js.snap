// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Offers Component should render a blank offers if no props are passed in 1`] = `
<div
  className="offers"
/>
`;
